doc/reference/trd-uart.md

```diff
@@ -0,0 +1,523 @@
+across all of them. Similarly, some services may need to be able to set
+the UART configuration but do not need to check it.
+
+Most devices using serial ports today use 8-bit data, but some older
+devices use more or fewer bits, and hardware implementations support
+this. If the data width of a UART is set to less than 8 bits, data is
+still partitioned into bytes, and the UART sends the least significant
+bits of each byte. Suppose a UART is configured to send 7-bit
+words. If a client sends 5 bytes, the UART will send 35 bits,
+transmitting the bottom 7 bits of each byte. The most significant bit
+of each byte is ignored.
+
+```rust
+pub enum StopBits {
+    One = 1,
+    Two = 2,
+}
+
+pub enum Parity {
+    None = 0,
+    Odd = 1,
+    Even = 2,
+}
+
+pub enum Width {
+    Six = 6,
+    Seven = 7,
+    Eight = 8,
+}
+
+pub struct Parameters {
+    pub baud_rate: u32, // baud rate in bit/s
+    pub width: Width,
+    pub parity: Parity,
+    pub stop_bits: StopBits,
+    pub hw_flow_control: bool,
+}
+
+pub trait Configuration {
+    fn get_baud_rate(&self) -> u32;
+    fn get_width(&self) -> Width;
+    fn get_parity(&self) -> Parity;
+    fn get_stop_bits(&self) -> StopBits;
+    fn get_flow_control(&self) -> bool;
+    fn get_configuration(&self) -> Configuration;
+}
+
+pub trait Configure {
+    fn set_baud_rate(&self, rate: u32) -> Result<u32, ErrorCode>;
+    fn set_width(&self, width: Width) -> Result<(), ErrorCode>;
+    fn set_parity(&self, parity: Parity) -> Result<(), ErrorCode>;
+    fn set_stop_bits(&self, stop: StopBits) -> Result<(), ErrorCode>;
+    fn set_flow_control(&self, on: bool) -> Result<(), ErrorCode>;
+    fn configure(&self, params: Parameters) -> Result<(), ErrorCode>;
+}
+```
+
+Methods in `Configure` can return the following error conditions:
+  - `OFF`: The underlying hardware is currently not available, perhaps
+    because it has not been initialized or in the case of a shared
+    hardware USART controller because it is set up for SPI.
+  - `INVAL`: Baud rate was set to 0.
+  - `ENOSUPPORT`: The underlying UART cannot satisfy this configuration.
+  - `FAIL`: Other failure condition.
+
+
+The UART may be unable to set the precise baud rate specified. For
+example, the UART may be driven off a fixed clock with integer
+prescalar. A call to `configure` MUST set the baud rate to the closest
+possible value to the `baud_rate` field of the `params` argument and a
+call to `set_baud_rate` MUST set the baud rate to the closest possible
+value to the `rate` argument. The `Ok` result of `set_baud_rate`
+includes the actual rate set, while an `Err(INVAL)` result means the
+requested rate is well outside the operating speed of the UART (e.g., 4MHz).
+
+
+3 `Transmit` and `TransmitClient`
+===============================
+
+The `Transmit` and `TransmitClient` traits allow a client to transmit
+bytes over the UART.
+
+```rust
+enum AbortResult {
+    Failure,
+    Success,
+}
```

Comment:
What is the purpose of including this enum instead of of just returning `Ok` or `Err(FAIL)`?

Reply:
The `Ok` vs. `Err` is about whether there will be a callback. If the return type is `Ok`, there will be a callback. If it's `Err`, there won't be (e.g., because no operation was oustanding). The `AbortResult` type is about whether the operation is actually aborted. So:
  - `Ok(AbortResult::Failure)` means there will be a callback and the operation wasn't aborted
  - `Ok(AbortResult::Success)` means there will be a callback and the operation was aborted
  - `Err(())` means there will not be a callback.

This trinary result really doesn't fit that well with `Result`, but I figured it was better to keep with linguistic support than roll a new type.

Reply:
Given that the Error in this case is just the unit type, this seems better suited to return `Option<AbortResult>`, or to just a custom type like you said. The `?` operator makes it easy to bubble errors up the call stack, but in this case we don't want callers to bubble up the error immediately, because one level higher than the caller it will not be clear that the `Err` does not indicate an actual error, but merely that there will not be a callback.

Reply:
Okay, I have read all of this now, and I think I have convinced myself that making abort split-phase is more confusing than helpful. The configure trait is already all synchronous, so we have implicitly discarded the possibility of an off-chip implementation requiring everything to be split. The HIL guidelines push towards split, but acknowledge pragmatism for synchronous calls.

Here, I put myself in the userspace position—cancelling operations is hard enough, but now I have to handle two separate asynchronous callbacks (the abort response and the original xmit response) in unknown order? What other case do we have multiple intertwined callbacks in flight for one driver?

Reply:
> I think what I am missing is why we need to generate a `transmit_abort` callback in the case of `AbortResult::Failure` -- there will still be the regular `transmitted_[word,buffer]` callback. Or perhaps equivalently, when `Err(())` would ever be generated in lieu of an `Ok(AbortResult::Failure)` (nothing to cancel?).
> 
> I'm also concerned about whether we have to enforce any ordering between the `abort` callback and the `transmitted_X` callback.
> 
> I think this is explained in more depth in the text below I didn't get to on first pass; will read that then circle back here.

There is no abort callback -- the callback it's talking about is the transmitted_{word,buffer} callback.

Reply:
> Given that the Error in this case is just the unit type, this seems better suited to return `Option<AbortResult>`, or to just a custom type like you said. The `?` operator makes it easy to bubble errors up the call stack, but in this case we don't want callers to bubble up the error immediately, because one level higher than the caller it will not be clear that the `Err` does not indicate an actual error, but merely that there will not be a callback.

The reasoning was less about bubbling up and more about trying to make it consistent with the transmit calls (`Ok` means callback, `Err` means no callback). But I agree that the benefits of this consistency are pretty tenuous. You are unlikely to have similar state transition code structures, because the call to cancel doesn't have a buffer it has to deal with in response to `Err`. 

I prefer the `Option` approach, mostly because I think the big deal is callback or no callback, not whether the operation was cancelled (you'll find that out later). So it's nice to be able to match on `Some`/`None` in the base case. Or it could be a type like

```rust 
enum AbortResult {
  Callback(bool),
  NoCallback,
}```